source/core/ExecutionContext.cpp

```diff
@@ -383,21 +384,37 @@ ExecutionState ExecutionContext::ExecuteSlices(Int32 numSlices, PlatformTickType
         }
     }
 
-    ExecutionState reply = kExecutionState_None;
+    Int32 reply = kExecSlices_ClumpsFinished;
     if (!_runQueue.IsEmpty()) {
-        reply = (ExecutionState) (reply | kExecutionState_ClumpsInRunQueue);
+        reply = kExecSlices_ClumpsInRunQueue;
     }
     if (_timer.AnythingWaiting()) {
-        reply = (ExecutionState) (reply | kExecutionState_ClumpsWaitingOnTime);
+        if (reply != kExecSlices_ClumpsInRunQueue) {
+            reply = kExecSlices_ClumpsWaiting;  // clumps waiting, but for less than 1 ms, we should be called again ASAP
+            Int64 timeToWait = gPlatform.Timer.TickCountToMilliseconds(_timer.NextWakeUpTime()-currentTime);
+            // This is the time of earliest scheduled clump to wake up; we return this time to allow the caller to sleep.
+            // Callers are allowed to call us earlier, say, if they set an occurrence to give us something to do.
+            // If we're called with nothing to run, we'll do nothing and return the remaining waiting time.
+            if (timeToWait < 0)
+                timeToWait = 0;
+            else if (timeToWait > kMaxExecWakeUpTime)
+                timeToWait = kMaxExecWakeUpTime;
+            // Negative return value (kExecSlices_ClumpsInRunQueue or kExecSlices_ClumpsWaiting) means we should be
+            // called again immediately/ASAP because there are clumps ready to run or will be in <1ms.
+            // Zero return (kExecSlices_ClumpsFinished) means the VI is completely finished
+            // and nothing is waiting to be scheduled.
+            if (timeToWait > 0)
+                reply = Int32(timeToWait);  // okay to truncate since kMaxExecWakeUpTime already checked range
+            // else time is < 1ms (after rounding from platform tick type), keep kExecSlices_ClumpsWaiting return value
+        }
     }
 #ifdef VIREO_SINGLE_GLOBAL_CONTEXT
     // TODO(PaulAustin): check global memory manager for allocation errors
 #else
     if (THREAD_TADM()->_totalAllocationFailures > 0) {
-        reply = kExecutionState_None;
+        reply = kExecSlices_ClumpsFinished;
     }
 #endif
-    _state = reply;
     return reply;
```

Comment:
Do you think it is worth asserting that if reply is < 0 that it is either -1 or -2?

Reply:
Well, not really.   I'm actually contemplating removing the distinction between those two values, I'm not sure there is a situation where a client would really need to distinguish between them, especially in JS where you can't wait < 4 ms, much less < 1 ms.

Reply:
Could it be helpful for with native embedders like raspberrypi? If so we could leave it as is. I don't feel strongly either way.

Reply:
What do you think about scrapping -2, and swapping the meaning of -1 and 0?
This way the return value doesn't feel like it has disjointed meaning between the timeout and the other cases.
0 really means no timeout, call me again immediately.
-1 is a traditional error or EOF code, meaning don't call me again.  Maybe could scrap the enum altogether.